core/compiler/src/main/scala/com/fortysevendeg/exercises/compiler/comments.scala

```diff
@@ -0,0 +1,341 @@
+import scala.tools.nsc.doc.base.comment._
+
+import scala.xml.NodeSeq
+import scala.xml.Xhtml
+
+import scalariform.formatter.{ ScalaFormatter }
+
+import cats.{ Id, Eq, Functor }
+import cats.data.Xor
+
+/** Facade for the different layers of comment processing. */
+object Comments {
+  import CommentZed._
+  import CommentParsing.ParseK
+
+  /** Type capturing the types for the name, description,
+    * and explanation fields.
+    */
+  type Mode = {
+    /** Name field type */
+    type N[A]
+    /** Description field type */
+    type D[A]
+    /** Explanation field type */
+    type E[A]
+  }
+
+  object Mode {
+    type Aux[N0[_], D0[_], E0[_]] = Mode {
+      type N[A] = N0[A]
+      type D[A] = D0[A]
+      type E[A] = E0[A]
+    }
+
+    /** Library comments. Required name. Required description.
+      * Require no description.
+      */
+    type Library = Aux[Id, Id, Empty]
+
+    /** Section comments. Require name. Optional description.
+      * Require no description.
+      */
+    type Section = Aux[Id, Option, Empty]
+
+    /** Exercise comments. Optional name. Optional description.
+      * Optional explanation.
+      */
+    type Exercise = Aux[Option, Option, Option]
+
+    // Isolated parse modes for specific fields.
+    // Each one ignores everything except the respective field.
+    // Note -- this is mainly used for testing.
+    type Name[A[_]] = Aux[A, Ignore, Ignore]
+    type Description[A[_]] = Aux[Ignore, A, Ignore]
+    type Explanation[A[_]] = Aux[Ignore, Ignore, A]
+  }
+
+  /** Helper function to parse and render a given comment. */
+  def parseAndRender[A <: Mode](comment: Comment)(
+    implicit
+    evPKN: ParseK[A#N],
+    evPKD: ParseK[A#D],
+    evPKE: ParseK[A#E],
+    evFD:  Functor[A#D],
+    evFE:  Functor[A#E]
+  ) =
+    CommentParsing.parse[A](comment).map(CommentRendering.render(_))
+
+}
+
+/* Our interface for interacting with Scaladoc's comment factory. */
+private[compiler] sealed trait CommentFactory[G <: Global] {
+  val global: G
+  def parse(comment: global.DocComment): Comment
+  // at the moment, this is provided for testing
+  def parse(comment: String): Comment
+}
+
+private[compiler] object CommentFactory {
+
+  /** Create a comment factory instance for a given global. */
+  def apply[G <: Global](g: G): CommentFactory[g.type] = {
+    new CommentFactoryBase with MemberLookupBase with CommentFactory[g.type] {
+      override val global: g.type = g
+      import global._
+      override def parse(comment: DocComment) = {
+        val nowarnings = settings.nowarn.value
+        settings.nowarn.value = true
+        try parseAtSymbol(comment.raw, comment.raw, comment.pos)
+        finally settings.nowarn.value = nowarnings
+      }
+      override def parse(comment: String) = parse(DocComment(comment))
+      override def internalLink(sym: Symbol, site: Symbol): Option[LinkTo] = None
+      override def chooseLink(links: List[LinkTo]): LinkTo = links.headOption.orNull
```

Comment:
Not sure of the implications of the possibility of a `null` flowing around, wouldn't `Option` be safer?